src/angular/18-item-total-pipe/order.component-solution.ts

```diff
@@ -72,35 +84,38 @@ export class OrderComponent implements OnInit, OnDestroy {
   }
 
   createOrderForm(): void {
-    this.orderForm = this.formBuilder.group({
-      restaurant: [this.restaurant?._id],
-      name: [null, Validators.required],
-      address: [null, Validators.required],
-      phone: [null, Validators.required],
+    this.orderForm = this.formBuilder.nonNullable.group({
+      restaurant: [this.restaurant?._id ?? '', Validators.required],
+      name: ['', Validators.required],
+      address: ['', Validators.required],
+      phone: ['', Validators.required],
       // PASSING OUR CUSTOM VALIDATION FUNCTION TO THIS FORM CONTROL
-      items: [[], minLengthArray(1)],
+      items: [[] as Item[], minLengthArray(1)],
     });
     this.onChanges();
   }
 
-  getChange(newItems: []): void {
-    this.orderForm?.get('items')?.patchValue(newItems);
+  getChange(newItems: Item[]): void {
+    this.orderForm?.controls.items.patchValue(newItems);
   }
 
   onChanges(): void {
     // WHEN THE ITEMS CHANGE WE WANT TO CALCULATE A NEW TOTAL
-    this.orderForm
-      ?.get('items')
-      ?.valueChanges.pipe(takeUntil(this.onDestroy$))
+    this.orderForm?.controls.items.valueChanges
+      .pipe(takeUntil(this.onDestroy$))
       .subscribe((val) => {
         this.orderTotal = this.itemTotal.transform(val);
       });
   }
 
   onSubmit(): void {
+    if (!this.orderForm?.valid) {
+      return;
+    }
+
     this.orderProcessing = true;
     this.orderService
-      .createOrder(this.orderForm?.value)
+      .createOrder(this.orderForm.getRawValue())
```

Comment:
I explained it before it is used in the training: https://github.com/bitovi/academy/pull/413/files#diff-34077455344a7b7ff8b62fe37140573fa5d99253b1f4991931159abb5206c704R136

> A FormGroup's `value` property is wrapped in `Partial` type because controls are removed from the form's value when disabled. For our case, we don't need to disable controls. We can use a FormGroup's `getRawValue()` method to access its value with the full type.